Allow items to be initialized lazily

ivoanjo opened the pull request:
This commits adds a `:lazy` option to items. A lazy item initialized with a proc (or block) will call the proc only to get the value once, and will then reuse the item on further calls.

The main usage of this is allowing a number of dependencies to be registered in the container without caring about their order (e.g. when a dependency depends on another), but still reusing the same objects, avoiding the need to instance them on each call to resolve, as is the default when not using lazy.

Hope this is acceptable, and thanks for your awesome work on dry-container!

ivoanjo committed:
Fix typo in documentation

ivoanjo committed:
Allow items to be initialized with a proc and then memoized

This commits adds a :memoize option to items. A memoized item
initialized with a proc (or block) will call the proc to get the value
only once, and will then reuse the item on further calls.

The main usage of this is allowing a number of dependencies to be
registered in the container without caring about their order (e.g. when
a dependency depends on another), but still reusing the same objects,
avoiding the need to instance them on each call to resolve, as is the
default when not using memoized.

AMHOL commented:
@ivoanjo thanks for the PR, could we call the option `memoize` instead of `lazy`? I think it's a better fit. Also we should probably raise an error if the `memoize` option is set to `true` and the proc parameters are not empty, WDYT?

solnic commented on lib/dry/container/item.rb:
```diff
@@ -4,22 +4,31 @@ class Container
       end
 
       def call
+        return lazy_item if options[:lazy] == true
```
It would be better to set `@lazy = options[:lazy]` in the constructor, add `attr_reader :lazy` and here just do `if lazy` this way we'd avoid hash lookup for each call

solnic commented on spec/support/shared_examples/container.rb:
```diff
@@ -185,6 +185,26 @@
+          expect(container.key?(:item)).to be true
+          expect(container.resolve(:item)).to eq('item')
+          expect(container[:item]).to eq('item')
+        end
```
This spec doesn't really check if it's the memoized item, as you create new string objects. It would be better to use a local var like `item = 'item'` and do `expect(container[:item]).to be(item)` (notice the `be` part, which checks object's identity rather than equality)

ivoanjo replied:
Yes indeed. I was using this spec only to test that the normal behavior was still supported, and was using the one after to test that the initialization was made only once, but I agree with your changes, as they tighten the spec.

ivoanjo replied:
Fixed in 1a2537082d9b7b9f7daad4a1845bbf50b4f79a35

solnic commented:
👍 for `memoize` instead of `lazy` + I left a couple of comments. Thanks for working on this btw, we've been thinking about this feature for a while, and it wasn't obvious if we need it or not, but at the end of day I think it's gonna be very useful.

I should mention that we added support for `Singleton` in dry-system (which uses this gem) but it's a limited solution, with this PR we'd have something more flexible. The only trick is to figure out how to make it work automagically with auto-registration in dry-system, but that's something to figure out later on...anyhow, just mentioning this as it's related :)

ivoanjo commented:
Wow, thanks for the awesome and quick feedback! Will definitely address the pointed suggestions/issues.

>  @ivoanjo thanks for the PR, could we call the option memoize instead of lazy? I think it's a better fit. 

I completely agree! I have to admit that maybe half the time I spent on this PR was thinking of how to name it. `singleton`, `lazy_singleton`, `initialize_once`, ... and I forgot `memoize`, which is obviously a much better name :smile: 

> I should mention that we added support for Singleton in dry-system (which uses this gem) but it's a limited solution, with this PR we'd have something more flexible. The only trick is to figure out how to make it work automagically with auto-registration in dry-system, but that's something to figure out later on...anyhow, just mentioning this as it's related :)

Cool! Indeed I'm working on a project that uses dry-container directly and has lots of stateless (function-like) objects --- I'm not a big fan of automatic registration, and prefer to state the dependencies explicitly, and was really missing this feature as otherwise we were having to precisely order dependencies which is a really big meh ;)

ivoanjo commented on spec/support/shared_examples/container.rb:
```diff
@@ -185,6 +185,26 @@
+
+      context 'with option lazy_singleton: true' do
+        it 'registers and resolves a proc' do
+          container.register(:item, proc { 'item' }, lazy_singleton: true)
```
Oops, just noticed I even ended up leaving one of my earlier names for this option here.

ivoanjo committed:
Raise error when initializing item with memoize: true but without a proc or block

ivoanjo committed:
Add expectation for memoized object identity

This test checks that the same object (the string 'item') is returned by
the container, not a new instance of String with the same contents.

ivoanjo committed:
Mark memoized_item as private

The underlying item should only be accessed via the `call` interface.

ivoanjo commented:
I've updated the PR with the feedback given, let me know what you think :smile:

AMHOL commented:
Awesome work, thanks @ivoanjo :)

:+1: Merging, will send an invite for repo access too

ivoanjo commented:
Thanks, and I'm honored for the repo access invite :)
